### ISSUE TITLE
Formal support for linking rlibs using a non-Rust linker

### ISSUE THREAD

**Issue by adetaylor** [rust-lang/rust#73632]
I'm working on a major existing C++ project which hopes to dip its toes into the Rusty waters. We:
  
- Use a non-Cargo build system with static dependency rules (and 40000+ targets)
- Sometimes build a single big binary; sometimes lots of shared objects, unit test executables, etc. - each containing various parts of our dependency tree.
- Perform final linking using an existing C++ toolchain (based on LLVM 11 as it happens)
- Want to have a few Rust components scattered throughout a very deep dependency tree, which may eventually roll up into one or multiple binaries

We can't:
- Switch from our existing linker to `rustc` for final linking. C++ is the boss in our codebase; we're not ready to make the commitment to put Rust in charge of our final linking.
- Create a Rust `staticlib` for each of our Rust components. This works if we're using Rust in only one place. For any binary containing several Rust components, there would be binary bloat and potentially violations of the one-definition-rule, by duplication of the Rust stdlib and any diamond dependencies.
- Create a single Rust `staticlib` containing all our Rust components, then link that into every binary. That monster static library would depend on many C++ symbols, which wouldn't be present in some circumstances.

We can either:
1. Create a Rust `staticlib` for each of our _output_ binaries, using `rustc` and an auto-generated `.rs` file containing lots of `extern crate` statements. Or,
2. Pass the `rlib` for each Rust component directly into the final C++ linking procedure.

The first approach is officially supported, but is hard because:
- We need to create a Rust `staticlib` as part of our _C++_ tool invocations. This is awkward in our build system. Our C++ targets don't keep track of Rust compiler flags (`--target`, etc.) and in general it just feels weird to be doing Rust stuff in C++ targets.
- Specifically, we need to invoke a Python wrapper script to consider invoking `rustc` to make a `staticlib` for every single one of our C++ link targets. For most of our targets (especially unit test targets) there will be no `rlibs` in their dependency tree, so it will be a no-op. But the presence of this wrapper script will make Rust adoption appear intrusive, and of course will have some small actual performance cost.
- For those link targets which _do_ include Rust code, we'll delay invocation of the main linker whilst we build a Rust static library.

The second approach is not officially supported. An `rlib` is an internal implementation format within Rust, and its only client is `rustc`. It is naughty to pass them directly into our own linker command line.

But it does, currently, work. It makes our build process much simpler and makes use of Rust less disruptive.

Because external toolchains are not expected to consume `rlib`s, some magic is required:
- The final C++ linker needs to pull in all the Rust stdlib `rlib`s, which would be easy apart from the fact they contain the symbol metadata hash in their names.
- We need to remap `__rust_alloc` to `__rdl_alloc` etc.

But obviously the bigger concern is that this is not a supported model, and Rust is free to break the `rlib` format at any moment.

Is there any appetite for making this a supported model for those with mixed C/C++/Rust codebases?

I'm assuming the answer may be 'no' because it would tie Rust's hands for future `rlib` format changes. But just in case: how's about the following steps?

1. [The Linkage section of the Rust reference](https://doc.rust-lang.org/reference/linkage.html) is enhanced to list the two _current_ strategies for linking C++ and Rust. Either:
   - Use `rustc` as the final linker; or
   - Build a Rust `staticlib` or `cdylib` then pass that to your existing final linker
(I think this would be worth explicitly explaining anyway, so unless anyone objects, I may raise a PR)
2. A new `rustc --print stdrlibs` (or similar) which will output the names of all the standard library rlibs (not just their directory, which is already possible with `target-libdir`)
3. Some kind of new `rustc` option which generates a `rust-dynamic-symbols.o` file (or similar) containing the codegen which is otherwise done by `rustc` at final link-time (e.g. symbols to call `__rdl_alloc` from `__rust_alloc`, etc.)
4. The Linkage section of the book is enhanced to list this as a third supported workflow. (You can use whatever linker you want, but make sure you link to `rust-dynamic-symbols.o` and everything output by `rustc --print stdrlibs`)
5. Somehow, we add some tests to ensure this workflow doesn't break.

A few related issues:
* #64191 wants to split the compile and link phases of rustc. This discussion has spawned from there.
* @dtolnay's marvellous https://github.com/dtolnay/cxx is not quite as optimal as it could be, because users can't use `-Wl,--start-group`, `-Wl,--end-group` on the linker line. (Per https://github.com/rust-lang/rust/issues/64191#issuecomment-629418541)
* the difficulties of using the `staticlib`-per-C++-target model happen to be magnified by #73047

@japaric @alexcrichton @retep998 @dtolnay I believe this may be the sort of thing you may wish to comment upon! I'm sure you'll come up with reasons why this is even harder than I already think. Thanks very much in advance.

**Comment by nagisa** [comment]
`rlib` is not a complete package you may need to link. It could very well just be a section containing just `rmeta` (serialized "metadata" representing checked Rust code) from which `rustc` would generate code at the time it is "forced" by generating a binary output (link/staticlib/cdylib). The fact that `rlib` contains any machine code at all is just an optimisation.

**Comment by petrochenkov** [comment]
I'm still don't understand what are the requirements exactly, but if rlibs (as they exist now) work for you, but staticlibs don't, then perhaps you need some variation of staticlibs with different bundling behavior for dependencies, rather than rlibs.

**Comment by adetaylor** [comment]
> The fact that `rlib` contains any machine code at all is just an optimisation.

Yep.

> What you really want is a `staticlib-nobundle` (not an actual thing) or something similar that would not bundle the the dependencies into produced `.a`.

I agree - this is exactly what we need.

**Comment by petrochenkov** [comment]
@adetaylor 
Could you elaborate a bit.
Suppose you have a Rust crate `root` with some tree of dependencies
```
  imm1 - dup
 /
root
 \ 
  imm2 - dup
   \
    indirect - native
```
where `imm1`, `imm2` and `dup` are other rust crates and `native` is some native dependency like C static library.

What the expected result of compiling `root` is supposed to be?

**Comment by adetaylor** [comment]
Right now, it would be a `staticlib` and that's fine.

The problem comes in this scenario:

```
  imm1 - dup
 /
root
 \ 
  imm2 - dup

root2
 \ 
  imm2 - dup

c_plus_plus_binary_a
 \ 
  root2 - [...]

c_plus_plus_binary_b
 \ 
  root1 - [...]

  root1 - [...]
 /
c_plus_plus_binary_c
 \ 
  root2 - [...]
```

In this case, we can't build `root` and `root2` as staticlibs, because they will conflict when we come to build `c_plus_plus_binary_c`.

**Comment by petrochenkov** [comment]
I don't mean right now, I mean what compilation of `root` and `root2` should produce to fit into your use case.

`libroot.a` + `libimm1.a` + `libimm2.a` + `libdup.a` + `libroot2.a` + some file in json or other machine-readable format describing the dependency trees?

**Comment by adetaylor** [comment]
Exactly! Except that the JSON file probably isn't necessary. Our build system (gn + ninja) works extremely hard to establish static dependency relationships in advance, so that a rebuild builds the minimal required set. We would expect to use those relationships in the subsequent linker invocation unless there were some reason that we needed to query it dynamically.

**Comment by alexcrichton** [comment]
This seems like a reasonable feature to me to support. The goal of rlibs was to reserve space for rustc to "do its thing" so we intentionally stabilize some format rather than accidentally do so. I don't think there's a need for a "staticlib-nobundle" because that's basically what an rlib is. While @nagisa is right in that rlibs can have varying contents, I think that it's fine to say for a use case like this you'd pass some flag to rustc saying "make sure I can pass this to the linker". Rustc, after all, goes to great lengths to pass rlibs directly to the linker to make linking fast.

I think this could work well perhaps with `-Z binary-dep-depinfo` (I forget the exact flag name). That'd basically tell you "here's all the other rlibs you need to link for this target" which would locate the standard library for you and other deps (which you'd probably ignore).

The gotchas that I could think of are:

* For allocators we'd probably just need to say "no custom allocator" or something similar. You'd probably run rustc with no arguments saying "generate me the allocator `*.a` file to pass to the linker", and that'd just run as part of your build system. That's also where we could shove anything else if it comes up in the future.

* You'll be on your own for deduplicating Rust dependencies across dynamic libraries. In your example of `c_plus_plus_binary_c` it's generally asking for trouble if for `root1` and `root2` are dynamic libraries and try to share dependencies. The compiler has internal support for this but you'd just be on your own. (this is totally fine, just wanted to point out!)

Overall I think we could get away with this by carefully designing how a "stable" pass-things-to-the-linker process would work. We could make sure to solve the constraints here while also giving ourselves some leeway so rustc can still change things in the future if needed. It seems worthwhile to me to integrate into other build systems like this, and it's definitely a pain point with rlibs since inception.

**Comment by nagisa** [comment]
> I think that it's fine to say for a use case like this you'd pass some flag to rustc saying "make sure I can pass this to the linker".

My suspicion is that the best and most holistic implementation of this kind of feature is represented by having a different crate-type alongside `rlib`. Implementing this functionality reliably can require a fair number of adjustments to how monomorphizations are generated, where they end up and what their visibility ends to being. I think the answer to these and many other similar such questions are meaningfully different enough for `rlib` and the hypothetical `libstatic-nobundle` that the two shouldn't be conflated.

**Comment by adetaylor** [comment]
Just to say: thanks all for the comments. I'm delighted that this doesn't seem like a completely idiotic request. From our (user's) point of view it is equally easy if this is a `crate-type=static-nobundle` as opposed to `crate-type=rlib`, either's fine.

We will bear in mind that this seems like a _plausible_ way forward as we continue to think about our linking strategies. We'll go quiet for a bit, but it's possible that in a few months we will pop up here and say "hi, we'd like to implement this, here's an RFC...". Watch this space. Obviously if anyone else gets there first that's even better :)

**Comment by bbjornse** [comment]
@adetaylor may I ask: have you found a workable solution? Or did you have to abandon Rust because of this issue?

In my experimentation I have used `--emit obj` to produce an object file, and create an archive from this (and other C++ object files part of the same library) using `ar`. I then use the system linker to produce the final artifacts. Is this not a viable solution?

**Comment by bjorn3** [comment]
> In my experimentation I have used --emit obj to produce an object file, and create an archive from this (and other C++ object files part of the same library) using ar. I then use the system linker to produce the final artifacts. Is this not a viable solution?

That doesn't work when using liballoc as the allocator shim is only generated when compiling a crate that needs to be linked and even then doesn't get included in the `--emit obj` object file, but a separate object file that gets deleted after linking. It can't be included in the main object file as `--crate-type lib --crate-type bin` in which case the allocator shim must be included for the bin, but must not be included for the lib.

**Comment by adetaylor** [comment]
@bbjornse We haven't abandoned Rust, but we also haven't made any real progress in this direction.

I know several other teams at various organizations are using the approach of building rlibs then passing them all into the final C++ linker. This still works. It still requires nasty hacks as described in this issue description, and could break at any time.

**Comment by bbjornse** [comment]
> That doesn't work when using liballoc as the allocator shim is only generated when compiling a crate that needs to be linked and even then doesn't get included in the --emit obj object file, but a separate object file that gets deleted after linking.

Interesting, thanks. Could you please explain: what exactly is the allocator shim, and why is it generated rather than part of a library? Does this relate to the `__rust_alloc` => `__rdl_alloc` remapping mentioned in the issue description?

Also, do you know if this is the only challenge with the object file approach? Are there other pieces which might be missing?

> @bbjornse We haven't abandoned Rust, but we also haven't made any real progress in this direction.
>
> I know several other teams at various organizations are using the approach of building rlibs then passing them all into the final C++ linker. This still works. It still requires nasty hacks as described in this issue description, and could break at any time

Thank you for your reply. This is somewhat disheartening. I'm still not abandoning hope for the `--emit obj` approach though: as far as I could tell the output was similar to that of the rlib, sans rmeta sections, and as you say using the rlibs directly can be made to work. Although passing rlibs directly to the native linker might never be officially supported, it sounds reasonable to me to expect that object files generated with `--emit obj` can be used in this way.

**Comment by bjorn3** [comment]
> Interesting, thanks. Could you please explain: what exactly is the allocator shim, and why is it generated rather than part of a library?

The allocator shim is what forward the allocator calls in liballoc to either a `#[global_allocator]` or the default allocator in libstd. Expanding it in place for `#[global_allocator]` won't work because if the crate with the `#[global_allocator]` and a crate linking to libstd get linked together the libstd default allocator wins as the dylib needs to have an allocator shim to link and the only allocator available when linking the dylib is libstd's default allocator.

> Does this relate to the __rust_alloc => __rdl_alloc remapping mentioned in the issue description?

Correct

> Also, do you know if this is the only challenge with the object file approach? Are there other pieces which might be missing?

Honestly I am not sure.

> Although passing rlibs directly to the native linker might never be officially supported, it sounds reasonable to me to expect that object files generated with --emit obj can be used in this way.

Neither rlibs, nor `--emit obj` contain the allocator shim. In fact they can't. For rlibs this is the reason mentioned earlier. For `--emit obj` this is because for `--crate-type lib --crate-type dylib` `--emit obj` will generate a single object file, but the rlib must not have the allocator shim and the dylib must have it.

**Comment by bjorn3** [comment]
I have been thinking about the rust compilation model lagely because of this and certain other deficiencies. No concrete actionable improvement yet though.

**Comment by bbjornse** [comment]
Thank you for your explanation. Have I understood correctly that
- If *any* crate used as part of a program or dynamic library uses `#[global_allocator]`, then that crate defines symbols `__rg_alloc` etc, and all references to `__rust_alloc` etc should be forwarded to `__rg_*` equivalents.
- Otherwise, all references to `__rust_alloc` etc should be forwarded to `__rdl_*` equivalents.
- No allocator shims are generated unless rustc is producing some kind of "final" artifact (`--crate-type {bin,dylib,cdylib,staticlib} --emit link`). In particular, they are not generated for `--crate-type lib --emit rlib,obj`.

I realize I don't understand why the rlib cannot include the shim. I see that it's only necessary when producing the final artifact, but I don't see how it would hurt to generate it immediately? (Unless I've misunderstood and `#[global_allocator]` is crate-local, of course.)

**Comment by bjorn3** [comment]
> I realize I don't understand why the rlib cannot include the shim. I see that it's only necessary when producing the final artifact, but I don't see how it would hurt to generate it immediately? (Unless I've misunderstood and #[global_allocator] is crate-local, of course.)

If you link to a dylib thaf already has an allocator shim, the rlib must not define an allocator shim, but the allocator shim of the dylib must be used. At the time of generating the rlib it is not yet known if it will be linked against a dylib. Only when compiling the final artifact.

**Comment by bbjornse** [comment]
I see, thanks. Is this only when the rlib shim would use the default allocator? Would it hurt to immediately generate the shim for crates which include a `#[global_allocator]`?

If crates require a shim at all, this means they use liballoc, right? And the prebuilt libstd already includes an allocator shim forwarding to the default allocator. How come this works, when another dylib might be providing a global allocator? Could a similar technique, linking with a "default allocator shim" dylib, work in a liballoc  +no-std setting?

**Comment by bjorn3** [comment]
When you dynamically link against libstd or any other dylib that uses liballoc, `#[global_allocator]` will be silently ignored. Instead the allocator shim embedder in said dylib will be used. This wouldn't be possible if rlibs that use `#[global_allocator]` immediately created the allocator shim.

**Comment by anforowicz** [comment]
### RE: --emit=obj

It seems to me that the [`--crate-type`](https://doc.rust-lang.org/rustc/command-line-arguments.html#--crate-type-a-list-of-types-of-crates-for-the-compiler-to-emit) switch controls the main output type of `rustc` (e.g. `rlib`, `staticlib`, `bin`, etc.) while the [`--emit`](https://doc.rust-lang.org/rustc/command-line-arguments.html#--emit-specifies-the-types-of-output-files-to-generate) switch can _optionally_ ask `rustc` to retain _additional_, _intermediate_ file types (e.g. `obj`, but also `mir`, `llvm-bc`, `dep-info`, etc.).  It feels that the `--emit` switch can be useful for debugging what is happening during compilation, but should not be depended upon for end-to-end results, since the intermediate/internal files produced when building an `rlib` crate may change in the future.  From this perspective, depending on the `--emit=obj` switch doesn't seem significantly different or less risky than depending on the details of what `--crate-type=rlib` produces.

### RE: monomorphization concerns

I see that @nagisa [has raised](https://github.com/rust-lang/rust/issues/73632#issuecomment-649162224) concerns that monomorphization within `rlib` might be fundamentally incompatible with the requirements here.  FWIW, I’ve tried to explore this a bit and didn’t find anything broken so far - please shout if I missed anything.

In particular, I tried building the following things:
- crate1: `pub fn sub_in_crate1<T: Sub<Output = T>>(a: T, b: T) -> T { a - b }`
- crate2a: `#[no_mangle] pub fn sub_from_7_in_crate2a(x: i32) -> i32 { crate1::sub_in_crate1(7, x) }`
- crate2b: `#[no_mangle] pub fn sub_from_42_in_crate2b(x: i32) -> i32 { crate1::sub_in_crate1(42, x) }`
- C++ executable that depends on crate2a and crate2b

There would be indeed 2 copies of the monomorphized `sub_in_crate1::<i32>`:
```bash
$ nm out/rel/rust_odr_test | grep crate1
00000000001f23d7 T _ZN20rust_odr_test_crate113sub_in_crate117h56a8a4a969227185E
00000000001f2422 T _ZN20rust_odr_test_crate113sub_in_crate117hf5a81c2f183f301cE
```

But this seems fine:
- The linker should hopefully be able to identify and eliminate the duplicated, identical binary code (and reduce binary bloat)
- The behavior of the compiled code seems unaffected:
    - I can’t see how having 2 copies might change the behavior of the compiled+linked code.  (e.g. I don’t see a way to introduce a singleton that becomes bifurcated/duplicated because of the above)
    - Rust seems okay with having duplicated monomorphisations in 2 rlibs when linking the dependent rlibs into the final executable.  (Maybe there is a deduplication step somewhere, but a cursory look didn’t find anything so far.)  And so it should be hopefully okay for C++ linker as well.

(More concrete and complete example is in a Chromium-specific code review tool: see [here](https://chromium-review.googlesource.com/c/chromium/src/+/3049552/4))

### RE: RFC

I guess the next step would be introducing an RFC (as suggested [above](https://github.com/rust-lang/rust/issues/73632#issuecomment-649162224)), but before doing that it is probably desirable to
- Make sure there are no issues with the general approach
- Try prototyping the changes (tweaking [`link_staticlib`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_codegen_ssa/back/link/fn.link_staticlib.html) to skip adding `each_linked_rlib` in presence of a new, hypothetical`-Z staticlib-with-no-transitive-deps` switch)

WDYT?

**Comment by jsgf** [comment]
> (--crate-type vs --emit)

One problem right now is that `--emit` is currently hashed into the crate hash, so, for example, `--emit metadata` and `--emit link` will generate different crate hashes for otherwise identical command lines. I'd really prefer this weren't the case, but I think this *particular* thing isn't relevant to this issue since by the time we get to linking crate hashes and rmeta should be irrelevant.

But aside from that, I'd agree that crate-type and emit have somewhat overlapping responsibilities (broadly, what kind of file(s) do I want?), but I'd characterize it as "crate-type sets the intrinsic properties of the output file" and "emit selects a number of auxillary outputs which can be cheaply generated along the way". But `--emit obj` and `--crate-type staticlib` are *very* close to each other really, ignoring the whole "and link in all the dependants" thing. (Does `--emit obj` also force codegen-units = 1, or is that just `--emit asm`?)

> 00000000001f23d7 T _ZN20rust_odr_test_crate113sub_in_crate117h56a8a4a969227185E
00000000001f2422 T _ZN20rust_odr_test_crate113sub_in_crate117hf5a81c2f183f301cE

I'm curious how this looks with `-Zsymbol-mangling-version=v0`, so that whatever's causing the difference isn't hidden in the hash.

I'm super interested in this effort. Currently Buck uses `staticlib` for C++ -> Rust dep edges, and I'd really prefer to have this be `rlib`/`staticlib-with-no-transitive-deps` and set up the dependencies to arrange all the other stuff appear on the final (C++) link line.

**Comment by bbjornse** [comment]
> (--crate-type vs --emit)

I agree that depending on `--emit obj`, without formal or documented guarantees about what the output will contain, feels fragile (like depending on linking with rlib). Case in point, when experimenting with this I found that `--emit obj` works differently with `--crate-type lib` and with `--crate-type staticlib`.

I don't feel that `--emit` is just a debugging option, though; rather, I understand it as an option to make `rustc` a powerful stand-alone tool usable in a non-Cargo context.

Reasons I still think `--emit obj` is a reasonable alternative:
- In contrast to rlib, it makes sense to me that the output from `--emit obj` is guaranteed to include object code. This is an improvement over rlib, where the presence of object code is an implementation detail (as noted [above](https://github.com/rust-lang/rust/issues/73632#issuecomment-647769750).)
- As @jsgf notes, it's `staticlib` without "link in all the dependants" - which is (almost) exactly what we want (i.e. an improvement over `--crate-type staticlib`)
- It makes it easy to create static libraries from a mix of C++ and Rust code. This is useful for bundling C++ bridges for a Rust library; bundling Rust bridges with a C++ library; or splitting a single library in Rust and C++ components.
- The capability to generate an object file is natural to expect when coming from a C or C++ world, and so is a useful feature in itself. In fact, it already *is* a feature in itself; it's just unclear to me at least how the output depends on `--crate-type`, and in what contexts the output will be usable.

**Comment by tmandry** [comment]
> Case in point, when experimenting with this I found that `--emit obj` works differently with `--crate-type lib` and with `--crate-type staticlib`.

Isn't this expected if `staticlib` links in all dependencies and `lib` doesn't? Or were there other differences?

**Comment by bjorn3** [comment]
`--emit obj` doesn't link in dependencies as it doesn't link at all. It just emits an intermediate object file that would be given to the linker otherwise.

**Comment by tmandry** [comment]
Ah. I was imagining a relocatable object file, like what `ld -r` produces, for `staticlib`. Not sure what behavior is actually desirable in that case.

**Comment by purpleposeidon** [comment]
I found the format of `--emit obj` to change w/ `--release`.

**Comment by bjorn3** [comment]
What do you mean with format? File format? If so from which file format did it change to which?

**Comment by bjorn3** [comment]
Oh, and are you using cargo or rustc directly? If cargo, there is no stable way to locate all object file necessary for linking as dependencies are not stored in a stable location.

**Comment by danakj** [comment]
Working on https://bugs.chromium.org/p/chromium/issues/detail?id=1296284 I have a better idea I think of what is needed or not needed here.

Requirements:
1. Linking a bunch of C++ .o files and Rust .o files together effectively, using the C++ frontend to run the linker.
1. We should _not_ link the Rust .o files together preemptively, they should be part of the final linking step as siblings to the C++ .o files. This avoids multiple copies of the rust stdlib and ensures linking decisions and optimization are happening in a single place. (This precludes `--crate-type=staticlib`.)
1. Crate compilation should behave the same as when producing an rlib - in particular it should not get slower or produce meaningfully different output.

**Current solution**: We use an .rlib file as a collection of .o files, and we must link it with `-Wl,--whole-archive`.

The proposal to use --emit=obj makes a lot of sense to me, it is a Rust .o file, which is what we need.

However, an .rlib is a collection of .o files because rustc splits the compilation up (via `--codegen-units`). So I am left with a question:

Does --emit-obj cause slower compilation by forcing `--codegen-units=1`? If so, that would contradict requirement #3.

Otherwise, --emit=obj producing different outputs is not necessarily a problem, they may not be meaningful. Perhaps they are artifacts of different codegen units for example. We'd need to understand what the differences are.

**Comment by tmandry** [comment]
We could easily make a mode that emits an archive of object files, I'm not sure what purpose `--emit=obj` has in _not_ doing that. Maybe the option predates having multiple codegen units, and now we need to preserve it for backwards compatibility?

**Comment by bbjornse** [comment]
My current understanding of this is that the requirements can be split in three:

1. Officially supported ability to generate object code for a crate, without its dependencies, into a file which can be passed to the system linker.
2. Link in libraries provided by the Rust toolchain.
3. Generate any required shims.

Looking at how each of these requirements are addressed by different alternatives:
1. I suspect this is already satisfied with `--emit obj`, with the only missing piece being a documented guarantee.

   But, the point about `--codegen-units` being forced to 1 is a valid concern. A quick search through rustc sources indicates this is true if we use `--emit obj=<dstfile>` - but not if we just use `--emit obj` (compiler/rustc_session/src/config.rs, function `should_override_cgus_and_disable_thinlto()`). Maybe someone knows if this is correct?
2. Support for linking rlibs is perhaps preferable as rlibs are actually distributed with the Rust toolchain. 
  Alternatively, maybe `library-archives` could be added as a `rustup` component containing `.a` equivalents to the shipped `.rlib`s. 
 A third option could be to add a new `mlib` type (multi-lib) or similar, working just like rlibs do today (i.e. both for linking and for rust compiltaion), and distribute those with the Rust toolchain.
3. Unaffected by the choice between `-emit=obj`, linking rlibs, or linking with the output of some new crate type `staticlib-nobundle`. The decision about what shim to use cannot be taken until link time, as only then do we know what shim should be used.

**Comment by bjorn3** [comment]
>  quick search through rustc sources indicates this is true if we use --emit obj=<dstfile> - but not if we just use --emit obj (compiler/rustc_session/src/config.rs, function should_override_cgus_and_disable_thinlto()). Maybe someone knows if this is correct?

https://github.com/rust-lang/rust/blob/4c55c8362de32be1530b2441c3e3a51e73edeb21/compiler/rustc_session/src/config.rs#L288 forces a single cgu for any `--emit obj` I think.

I wonder if instead of allowing rlibs to be linked outside of rustc we could allow C object files to be linked with rustc, so rustc takes the place of the linker, handling any orchestration for rust libraries and then forwarding to the system linker or any other linker provided with `-Clinker`.

**Comment by bbjornse** [comment]
> forces a single cgu for any `--emit obj` I think

I see this function where `false` is returned is named `is_compatible_with_codegen_units_and_single_output_file()` - and the code using it at line 1537 seems to only look at emit-flags with an explicit path:
```rust
    let incompatible: Vec<_> = output_types
        .0
        .iter()
        .map(|ot_path| ot_path.0) // This one - only proceed if we actually have an explicit path?
        .filter(|ot| !ot.is_compatible_with_codegen_units_and_single_output_file())
        .map(|ot| ot.shorthand())
        .collect();
```
Which is what makes me suspect that it only forces cgu=1 with an explicit path?

**Comment by bbjornse** [comment]
> I wonder if instead of allowing rlibs to be linked outside of rustc we could allow C object files to be linked with rustc, so rustc takes the place of the linker, handling any orchestration for rust libraries and then forwarding to the system linker or any other linker provided with `-Clinker`.

I suspect this is a poorer fit for C++ codebases wanting to integrate Rust code into what they have. The link setup would then have to be recreated *exactly* as it were before, but indirectly through rustc. This is more complex both to read and write than just writing the desired linker invocation directly. It is also more fragile, as rustc must be trusted to produce the exact same linker invocation across rustc versions (sans differences in rust library names), and of course across runs.

**Comment by bjorn3** [comment]
> Which is what makes me suspect that it only forces cgu=1 with an explicit path?

`ot_path` here has the type `(OutputType, Option<PathBuf>)` I think. `None` here represents the case where no path is given. That line you mentioned ignores the `Option<PathBuf>` and only returns the `OutputType`.

**Comment by jsgf** [comment]
Semantically `--emit obj` seems fine to me, but the codegen units = 1 problem would be a real practical blocker. Something that emits a plain .a file would be a fine replacement and seems straightforwardly compatible with multiple cgus. In practice this could be a plain .rlib, or an rlib with an extra step of stripping out the rmeta files if that confuses the linker.

Ideally this would also support thin archives in environments which have such a concept, to avoid some needless file copies (which I think is an orthogonal potential improvement for rlibs). 

@bjorn3:
> I wonder if instead of allowing rlibs to be linked outside of rustc we could allow C object files to be linked with rustc, so rustc takes the place of the linker, handling any orchestration for rust libraries and then forwarding to the system linker or any other linker provided with -Clinker.

I don't think that would work in practice. For the cases I'm interested in, the linking step is immensely complex with lots of variants, the amount of code is enormous and mostly C++, with Rust comprising probably a fraction of a percent. It would be very hard to justify reengineering the build process to do linking via `rustc`. (The only way I can see that happening is if rustc had a wrapper to allow it to be a drop-in replacement for the existing linker. But if that were possible, I'm not sure what the benefit would be since it wouldn't be able to do any Rust-specific things.)

For the cases where the top-level binary is Rust, then we do use `rustc` as the linker, and inject all the non-Rust objects via `-Clink-arg` etc so they can be consumed by the underlying linker.

Somewhat unrelated aside: What's the distinction between `--emit` and `--crate-type`? For the most part I think the options make sense to me, but using `--crate-type` for `cdylib`/`staticlib` seems like an exception; they seem much closer match for `--emit`. But I don't have a crisp mental model for distinguishing when which option is the most appropriate. `--crate-type` for "binary" vs "library" makes sense, and `--emit` for "deps"/"asm" seem reasonable. But the middle ground of "rlib" vs "dylib" vs the native link variants seem like they could also be emit. And then there's the problem that `--emit metadata`'s meaning changes depending on what other emit options you have.

**Comment by bbjornse** [comment]
> That line you mentioned ignores the Option<PathBuf> and only returns the OutputType.

Ah, you are right of course. Feeling stupid now :slightly_smiling_face: 

Not giving up though, I'll suggest one more way in which the `is_compatible_with_codegen_units_and_single_output_file()` function can live up to its name. Looking at the code using the list of incompatible output types (shortened a bit):
```rust
    if !incompatible.is_empty() {
        match codegen_units {
            Some(n) if n > 1 => {
                if matches.opt_present("o") {
                    // <...>
                    codegen_units = Some(1);
                }  // ELSE don't touch codegen_units?
            }
            _ => {
                codegen_units = Some(1);
            }
    }
```
I.e. it appears cgu is set to 1 only if the `-o` option is used, or if it was not set. Testing this:
```bash
$ echo 'pub mod mod1 { pub fn f() {} } pub mod mod2 { pub fn f() {} }' | rustc --crate-type lib --emit obj -C codegen-units=2 --out-dir=/tmp/example -
warning: ignoring emit path because multiple .o files were produced

warning: 1 warning emitted                                                                                                                                                                                                
$ ls /tmp/example
rust_out.rust_out.b25ea2e0-cgu.0.rcgu.o  rust_out.rust_out.b25ea2e0-cgu.1.rcgu.o                                                                                                                                                                                                    
$ 
```

So it would appear we can get codegen-units > 1 with `--emit obj`, albeit with a warning?

Of course, a mode for emitting archives as @tmandry suggested would in any case be preferable  :slightly_smiling_face:

**Comment by jsgf** [comment]
> So it would appear we can get codegen-units > 1 with --emit obj, albeit with a warning?

It's a bit awkward if we don't know what their filenames are though. Ideally it's something we could know in advance (eg by specifying the, or they're computable in a straightforward way). I guess we could use artifact notifications or just inspect what was left in a previously empty directory after the fact.

**Comment by danakj** [comment]
I came across an issue with using an archive output. The linker drops object files from an archive when there is no unresolved symbol that the object file satisfies. In particular, this means that in a test binary, a file of unit tests which are registered with static initializers produces object files that get dropped if linked through an archive. This happens because the object files have no external calls into them, creating unresolved symbols. This gets especially interesting in Rust where a .rs file can be split into many object files, so even hacks to call some method in each .rs file do not resolve it.

We are working around this while we're linking .rlib files by passing --whole-archive to the linker for the rlibs. However not all platforms support --whole-archive. The AIX linker for example does not: https://chromium-review.googlesource.com/c/chromium/src/+/3470980/ 

This means we need to either:
1. Unarchive the object files from rustc's emitted archive.
2. Have rustc generate object files directly, and emit a list of them (to stdout?)

But I've also noticed that [Files cannot be extracted from a thin archive](https://www.systutorials.com/docs/linux/man/1-ar/), which was proposed earlier https://github.com/rust-lang/rust/issues/73632#issuecomment-1038604003.

**Comment by bjorn3** [comment]
Static initializers don't work reliably even when using rustc as linker either. I think you probably shouldn't use static initializers with rust for the time being until it is fixed.

**Comment by bjorn3** [comment]
Native linking modifiers fixes it for C static libraries. They don't apply to regular rust dependencies.

**Comment by jsgf** [comment]
@hlopko 
> [thin archives]

Yeah, I think you're right. From my perspective, thin archives are mostly a disk IO/copy bandwidth saving rather than specifically for disk space saving, though that's a nice side benefit. If the flow is cc -> .o -> ar -> .a, then it can get expensive if you have, for example, lots of debug info enabled (typically 10x the .text size). But if you're right that llvm directly emits .a without an intermediate .o (and I think you are), then that's moot.

`--start-lib`/`--end-lib` does give similar benefits to thin archives, but unfortunately they're incompatible with link groups (--start-group/--end-group).  Though that's only relevant for `gold` since `lld` effectively acts as if the entire link line were in a link group.

And for the record, my primary interest is building with [Buck](https://buck.build) which shares many characteristics with Bazel.

An additional constraint with all this is that we need to have explicit dependencies on all the sysroot crates, including all their dependencies, expressed as `prebuild_rust_library` rules (for now, though we're also looking into building libstd from source with Buck rules). And in addition, prevent rustc from implicitly using sysroot. (This isn't strictly necessary for pure rust builds, but it is to get the right sysroot libraries to a final C++ linker.)

This turns out to be surprisingly hard - not the dependencies themselves, but working out when a crate has a dependency on a sysroot crate.  (I have a little python script which can generate sysroot deps from the crate metadata itself.)

We can generally assume everything depends on `std` (or `core` for `#![no_std]` crates). This is reasonably simple (but awkward because you can't identify no_std crates from the metadata alone).  But it seems you can *also* define a dependency on `rustc-std-workspace-(std|core|alloc)` or `compiler_builtins` in a Cargo.toml, which cargo recognizes and basically suppresses on the assumption that rustc will just pluck that dependency from its sysroot. We must suppress sysroot access, and replace it with an explicit dependency on the appropriate sysroot rule targets.

This seems reasonably easy to do in an ad-hoc way for small examples, but I've yet to make it work for a full-scale build using a non-trivial number of third-party crates.io deps.

**Comment by jsgf** [comment]
> No, the rustc-std-workspace-* crates are in crates.io with an empty content. In the rust build system there is a [patch.crates-io] section replacing them with alternative versions which re-export the respective standard library crate. The compiler-builtins version on crates.io is the exact same used by the standard library. (in fact the standard library gets it from crates.io)

I think I see - I'll need to dig into it some more. But I did notice that Cargo has specific knowledge of the rustc-std-workspace* names baked into it, and they don't appear as normal resolved dependencies in the `cargo metadata` output. This led me to believe that they're intended to be resolved against whatever sysroot your rustc is using (since quite a few crates.io packages depend on them, I guess as a result of being used within sysroot).

**Comment by bjorn3** [comment]
> But I did notice that Cargo has specific knowledge of the rustc-std-workspace* names baked into it

The only occurences of rustc-std-workspace-core in the cargo source code are for tests and the implementation of `-Zbuild-std`. It doesn't have any special meaning for regular compilation. https://github.com/rust-lang/cargo/search?q=rustc-std-workspace-core

The patching of rustc-std-workspace-* happens at https://github.com/rust-lang/rust/blob/1103d2e914b67c18b0deb86073c26c6aefda761d/Cargo.toml#L128-L132

The rustc-std-workspace-core and rustc-std-workspace-alloc on crates.io is empty. Docs.rs doesn't render empty files correctly it seems: https://docs.rs/crate/rustc-std-workspace-core/latest/source/src/lib.rs The rustc-std-workspace-std on crates.io re-exports libstd from the sysroot: https://docs.rs/crate/rustc-std-workspace-std/latest/source/src/lib.rs

The patched versions in the rust repo depend on the respective crate and re-export it:

https://github.com/rust-lang/rust/blob/1103d2e914b67c18b0deb86073c26c6aefda761d/library/rustc-std-workspace-core/Cargo.toml#L13-L14

https://github.com/rust-lang/rust/blob/1103d2e914b67c18b0deb86073c26c6aefda761d/library/rustc-std-workspace-core/lib.rs#L1-L2

When compiling the standard library it is not possible to depend on any crate being in the sysroot as the sysroot is empty at that point. It is only filled after building the standard library.

**Comment by danakj** [comment]
I linked https://github.com/rust-lang/rust/issues/94138#issuecomment-1048077884 here. The summary from there is that I've now hit my first case of a linker error due to linking rlibs. And the error does not occur if you link the --emit=obj object file instead.

**Comment by mhammerly** [comment]
On the topic of allocator shims: I took a crack at implementing a `-C force-allocator-shim` per @jsgf's suggestion elsewhere and when I was 90% finished I saw https://github.com/rust-lang/rust/pull/86844 "Support #[global_allocator] without the allocator shim" from @bjorn3 who has been active here. #86844 has been open since July so I put my PR up to see if it'd bump the discussion.

I think either PR can be used to work around the allocator shim issue. I think #86844 would work with a single-`.o` `--emit=obj` strategy (right?) while #94389 behaves like a normal allocator shim with a separate `.o` in the `.rlib`. If I run my local #94389 `rustc` with `--emit=obj` the resulting `.o` has no allocator symbols in it.

Archives seem nicer than `--emit=obj` to me, but I'm not really equipped to argue either way.

**Comment by pcwalton** [comment]
For those interested, I have a minimal test case to demonstrate the duplicate symbols problem here: https://github.com/pcwalton/rust-staticlib-duplicate-symbols-test

The problem can be triggered with just two crates, plus a C program that links them. Crate `a` exports symbols `foo` (mangled) and `bar` (unmangled, so C can call it). Crate `b` links against `a` *as an rlib* (i.e. in Rust) and exports a symbol `baz`(unmangled) which calls `foo` internally. The C program `c` links against `a` and `b` *as staticlibs, in that order* and calls `bar` and `baz`.

Here's how the linking works. The linker first goes to find `bar` and resolves it to the object file inside `liba`. Then it looks up `baz` and resolves it to the object file inside `libb`. The linker then has to resolve the call to `foo` inside `baz`, but it can't use the copy inside `liba` because `rustc` gave it a different symbol name. So the linker goes to find `foo` inside `libb`, but the object file containing `libb`'s copy of `foo` also contains a definition for `bar`. This causes a multiple definition error, and `ld` fails to link.

Note that whether you get a linking error is sensitive to many factors. If you pass `-lb -la` instead of `-la -lb`, you get no error, because the linker will find both `bar` and `baz` inside `libb` and won't look inside `liba` at all. If you remove the call to `foo` inside the definition of `baz`, you get no error, because `ld` won't look at the object file inside `libb` containing the duplicate copy of `baz`. There are even cases (not this one) in which changing the relative alphabetical ordering of symbols will cause the linker to resolve names in different orders and can make the difference between linking succeeding and failing.

**Comment by pcwalton** [comment]
> Semantically `--emit obj` seems fine to me, but the codegen units = 1 problem would be a real practical blocker. Something that emits a plain .a file would be a fine replacement and seems straightforwardly compatible with multiple cgus. In practice this could be a plain .rlib, or an rlib with an extra step of stripping out the rmeta files if that confuses the linker.

There is a flag you can pass to `ld`—`ld -r`—that should allow us to continue to use codegen units with `--emit obj`, by allowing separate object files to be combined into a single relocatable one for subsequent linking. The RFC that comes out of this should probably mention `ld -r` as a potential alternative solution. That being said, I don't love `ld -r` simply because in my experience it's not that well tested and can have issues. I also don't know if Windows has such a thing.

**Comment by pcwalton** [comment]
What would folks here think if I wrote a pre-RFC for a minimal feature that stabilizes some of what rustc does today with rlibs? Effectively this would just state that rlibs are archives in the native system format, containing object files in the native system format, plus some extra `.rmeta` files, that comprise all the symbols in the crate, but not any upstream dependencies. The RFC would *not* specify the format of the metadata—that would be considered perma-unstable and subject to change. Effectively this RFC would simply be a minimal commitment that rustc will not move away from emitting native system archives if `--crate-type=rlib` is specifically passed to the compiler.

**Comment by bjorn3** [comment]
For some targets we may need to invent our own format if there exists no standard archive format. For example rust-gpu uses tar for it's spirv target. Furthermore there may be performance benefits to using different archive formats or even emitting a single object file containing both code and crate metadata in case only a single cgu is used. Also without knowing where to find libcore and other parts of the standard libraries rlibs are useless. We would have to add some way to know which crates an rlib needs to link to. And finally it would permamently close the door for things like mir-only rlibs or similar things where codegen would happen more lazily to improve compiler performance.

**Comment by pcwalton** [comment]
> For some targets we may need to invent our own format if there exists no standard archive format. 

This is fine. I only care about existing architectures.

> Furthermore there may be performance benefits to using different archive formats or even emitting a single object file containing both code and crate metadata in case only a single cgu is used.

Note that I only suggest stabilizing `--crate-type=rlib` when *explicitly* specified as such. We could introduce new crate types if needed, and switch them to be the default.

> Also without knowing where to find libcore and other parts of the standard libraries rlibs are useless.

That's not a problem in practice; our build system knows where to find them.

> We would have to add some way to know which crates an rlib needs to link to.

No, we wouldn't. This feature is intended for non-Rust build systems, which already know all that information.

> And finally it would permamently close the door for things like mir-only rlibs or similar things where codegen would happen more lazily to improve compiler performance.

See above; I'm only suggesting stabilizing the contents of explicit `--crate-type=rlib`. Nothing in this RFC precludes the possibility of introducing `--crate-type=mir-rlib` or `--crate-type=rlib2` or something like that.

Can you explain what solution you would prefer for our use case? We cannot do any of (a) using `rustc` for final link; (b) avoiding diamond dependencies; (c) sacrificing parallelism by forbidding multiple CGUs per crate.

**Comment by bjorn3** [comment]
> That's not a problem in practice; our build system knows where to find them.

How does it now for sysroot crates without depending on unstable implementation details which kind of defeats the point of stabilizing this.

> No, we wouldn't. This feature is intended for non-Rust build systems, which already know all that information.

They can't know which parts of the standard library are used unpess rustc tells it. You can't simply add all individual crates in the `rustlib` directory for the target as that may include unused crates or even crates which may only be linked in specific cases. (for example only one of panic_abort and panic_unwind may be linked and one of them must be linked if libstd is used. their existence is not guaranteed though. we may switch to something similar to the allocator shim)

> See above; I'm only suggesting stabilizing the contents of explicit --crate-type=rlib. Nothing in this RFC precludes the possibility of introducing --crate-type=mir-rlib or --crate-type=rlib2 or something like that, and making a future rlib2 the default.

Such a feature would likely require every crate to be compiled with it. For as long as we don't have   stable -Zbuild-std that means the standard library would have to be compiled with this feature, but that would break build systems depending on rlibs containing useful object files.

> Can you explain what solution you would prefer for our use case? We cannot do any of (a) using rustc for final link; (b) avoiding diamond dependencies; (c) sacrificing parallelism by forbidding multiple CGUs per crate.

Have rustc tell you exactly what you need to link and have this rustc invocation generate any necessary support files like the allocator shim.

On a side note: The current compilation model of rustc hasn't been well thought out. It consists of a bunch of features bolted on top of each other creating a bit of a mess. I feel like we should first document the current compilation model and have a unifying vision for a better compilation model before we can think about how to incrementally improve (or maybe even rewrite from scratch) the compilation model of rustc to prevent painting ourself into a corner with suboptimal compilation performance while at the same time having bad integration with the C compilation model and permanently bad dylib support.

**Comment by bjorn3** [comment]
By the way I forgot if I asked this already, but would a rustc command to bundle multiple rlibs into a single staticlib without requiring any rust source for this staticlib work? So something like `rustc --crate-type staticlib -Zlink-libs liba.rlib libb.rlib -o libcombined.a`. This should allow diamond dependencies I think. You could run it just before invoking the system linker.

**Comment by pcwalton** [comment]
That may work as long as rustc isn't going to magically read any files that the build system doesn't know about. Keep in mind that many large-scale build systems are distributed and so the tool has to precisely know all of the dependent files that rustc is going to look at before invoking the compiler.

I'll let @jsgf comment on whether this works for Buck specifically, since I'm not as familiar with its inner workings.

**Comment by mhammerly** [comment]
@bjorn3 Just making sure I understand your suggestion correctly: before the final link, the non-Rust build system (GN + Ninja, Buck, Bazel, whatever) pulls a flat list of Rust dependencies out of the dependency graph, runs your `rustc -Zlink-libs` command, and uses the output in the link line?

If that's right, I think a cruder version of that idea was covered in the OP:
> Create a Rust staticlib for each of our output binaries, using rustc and an auto-generated .rs file containing lots of extern crate statements. Or,

I'm not positive but I think the downsides listed in the OP would still apply to your idea too. Propagating flags like `--target` or `-Z link-native-libraries=no` to the Rust pre-link would be difficult, and I think the build speed impact on a large project of basically double-linking all the Rust code would be rough. Those points were written about gn + ninja, but I think they'd apply to Buck as well from my (limited) experience.

I can't speak to issues with compiler performance, but I think the way to improve support for big multi-language projects is to make `rustc`'s observable behavior less special:
- generate an "exportable" unbundled static archive (what @pcwalton is offering to advance here)
- generate an unbundled shared library without `dylib`'s baggage (so: skip the allocator shim, and defer errors about missing panic or alloc error handlers to the "boss toolchain")
  - this is a separate rabbit hole, just including it for completion

Granted those, there are some gaps we need official solutions for:
- generating an allocator implementation since `rustc` won't do it for us (#86844)
  -  from @bjorn3's comment we may need this for `panic_*` later too
- providing the standard library to the linker since `rustc` is no longer bundling it
  - users relying on implicit deps on the stock sysroot could expand OP's described `rustc --print stdrlibs` on the link line
  - or, users can pass `--sysroot /dev/null` and provide explicit sysroot deps. "Unused deps" checks can be handled with #96067

For what it's worth, the unbundled static archive + explicit sysroot deps approach is working fine at Meta. I don't think we're that far from serviceable official support, but I'm sure I'd learn a lot from an RFC discussion.

> I feel like we should first document the current compilation model and have a unifying vision for a better compilation model before we can think about how to incrementally improve (or maybe even rewrite from scratch) the compilation model of rustc to prevent painting ourself into a corner with suboptimal compilation performance while at the same time having bad integration with the C compilation model and permanently bad dylib support.

How can we figure out whether beginning this revamp is worth prioritizing soon? The pre-RFC @pcwalton offered to write (or a full RFC) feels like a good way to get wider input on whether that's necessary.

**Comment by bjorn3** [comment]
> Just making sure I understand your suggestion correctly: before the final link, the non-Rust build system (GN + Ninja, Buck, Bazel, whatever) pulls a flat list of Rust dependencies out of the dependency graph, runs your rustc -Zlink-libs command, and uses the output in the link line?

Indeed

> and I think the build speed impact on a large project of basically double-linking all the Rust code would be rough.

Thin archives would make it a simple listing of all files and generating any necessary auxiliary object files rather than copying everything I think.

**Comment by adetaylor** [comment]
> Just making sure I understand your suggestion correctly: before the final link, the non-Rust build system (GN + Ninja, Buck, Bazel, whatever) pulls a flat list of Rust dependencies out of the dependency graph, runs your rustc -Zlink-libs command, and uses the output in the link line?

For what it's worth, this would make it potentially harder for us to introduce Rust into our project. Rust code may appear anywhere in our huge and complex dependency graph, eventually propagating to some final C++ linker target. Our build system discourages global knowledge by design (for speed/scalability reasons) so there's no way to know in advance whether a build target would have Rust anywhere in its dependency tree. We'd have to at least _consider_ running `rustc -Zlink-libs` for _every_ C++ linker invocation. This would have a small but non-zero impact on the build times of pure-C++ parts of our codebase, which would be difficult for us to justify at this point.

**Comment by pcwalton** [comment]
Based on feedback I'd like to revise my proposal to introduce a new crate type, `--crate-type=rlib0`, which is simply "what Rust calls an rlib today". This is intended for use with external build systems that need to be able to count on handling rlibs. The reason for introducing a new crate type is so that `--crate-type=rlib` can change in the future without breaking these external build systems.

**Comment by jsgf** [comment]
Woah, lots of activity here.

@bjorn3, I think it's worth restating what we're trying to achieve here, and what the environment and constraints are. I'll speak for @pcwalton and @mhammerly; I can't speak to @adetaylor's environment, but what he says resonates with me and I think there's a lot of commonality here.

So, we're talking about Rust interop with other languages. C++ is the primary one, but we're also interested in at least Python, Java, Ocaml, Haskell. Rust is very appealing in this role, because its execution model is very similar to C/C++ - there's no real operational difference between running a Rust binary and a C++ one from a deployment or monitoring point of view.

We would like to extend this to the compilation and linking point of view - that is, that a given piece of Rust code which exposes an API with a C ABI, we would like to be able to freely integrate that into a build while localizing the knowledge that it's implemented in Rust.

Note that while I'm mostly going to be addressing this in the context of "big builds out of a monorepo", I see all this as solving a much more commonplace problem. We're seeing an increasing number of projects where C/C++ codebases are being incrementally rewritten in Rust (eg [curl](https://www.abetterinternet.org/post/memory-safe-curl/)), so we'll start seeing more and more hybrid Rust/C/C++ build and link cases at all scales. The more that we can make Rust Just Work in all kinds of build systems when interoperating with C, easier it will be to justify RIIR projects. But this means trying to avoid things like special link time steps.

We have very large dependency graphs (as @adetaylor mentions), with arbitrary dependency edges between Rust and C++, in both directions. The top-level might be Rust, C++ or some other language which can bind with a C/C++ ABI.
 A single top-level dependency (eg executable but not necessarily) could have tens of thousands of transitive dependencies, reachable via hundreds of thousands of dependency paths - that is, a single Rust library could be depended upon via many paths, both via Rust and C++ (`std` or `core` being the obvious maximal cases, as every other crate will have a dependency edge to them).

(Also related: We have quite a lot of technologies based on processing object files and post-executables, eg [BOLT](https://research.facebook.com/publications/bolt-a-practical-binary-optimizer-for-data-centers-and-beyond/) - so having everything use a uniform representation helps a lot there too. We'd like Rust code to also benefit from those optimizations minimal special effort.)

The goal here is to make sure that *every crate is compiled precisely once*, regardless of how many times its depended upon, and how many dependency paths there are.

While the fundamental problem with `staticlib` is that it bundles all the transitive dependencies and causes massive symbol duplication (linker errors) and inefficiency (copying) in this compilation environment, it also shares a problem with any proposal which has a different crate type or compilation mode for "Rust for consumption by C++" vs "Rust for consumption by Rust" - it involves building the crate multiple times.

As a result, I've come to the conclusion that we need a single crate type and file format which is suitable for both C++ and Rust consumption, which is the essence of @pcwalton's proposal - rlib, as it currently stands, is actually fine for this.

(Though I'd probably prefer something like `--crate-type rlib --emit universal-link=...`.)

To address some of the specifics:

> For some targets we may need to invent our own format if there exists no standard archive format. For example rust-gpu uses tar for it's spirv target.

We're only concerned with language interop, so we would only be concerned with using the same archive or object file format that other languages use in that environment. If there are no other languages or pre-existing formats, then Rust would be free to invent new things.

> Furthermore there may be performance benefits to using different archive formats or even emitting a single object file containing both code and crate metadata in case only a single cgu is used.

If that's the case, then using a single object file with the rmeta embedded as a section (or similar) would be fine, and should be completely compatible with the platform linking model.

> Also without knowing where to find libcore and other parts of the standard libraries rlibs are useless. We would have to add some way to know which crates an rlib needs to link to. 
> How does it now for sysroot crates without depending on unstable implementation details which kind of defeats the point of stabilizing this.

Yes, that's a big issue - I'll go into it a bit more below. @pcwalton's proposal is necessary but far from sufficient - there are a lot of other details to sort out. But this is a starting point.

> And finally it would permamently close the door for things like mir-only rlibs or similar things where codegen would happen more lazily to improve compiler performance.

Not necessarily, but it does make things more complex (because that's intrinsically complex in a polygot build environment). In the polyglot dependency graph I describe above, any point where you have a non-Rust -> Rust dependency would require the crate content to be something that the other language can reference (ie object code, or a proxy like llvm-bc for LTO). Within the connected subgraphs of Rust->Rust dependencies you could use Rust-specific representations, but that's still complex without analyzing the entire dependency graph (which we'd like to avoid if possible).

## sysroot dependencies

There's a few problems here:
1. How do we know which parts of sysroot a given crate requires?
2. What are the dependencies *between* sysroot crates?
3. How do we handle non-local configuration like allocators and panic handlers?

Fundamentally 1 is awkward because everything in sysroot is implicitly available to all crates, and there's no current requirement to specify any dependencies/requirements explicitly. (Proc-macros being a partial exception, by happenstance.) Distinguishing `std` from `core` users is particularly awkward, because the signal for that is buried inside the crate's source rather than specified externally (eg passed in as a `rustc` command-line argument). Likewise `alloc` etc are Just There. To solve this we just specify all the likely sysroot crates as dependencies for everything (incl `std` and `core` in the prelude, `alloc` and `compiler_builtins` as `noprelude`). An unfortunate consequence is that everything gets `std` regardless of whether it needs is, which is fine for the server apps I'm mostly interested in, but it doesn't work for mobile/embedded use-cases. (It's also really awkward that 

To solve 2, I currently have a hacky script which, given a list of "root" sysroot rlibs (`std`, `core`, `alloc`, and the `rustc_std_workplace_` versions), it uses `rustc -Zls` to walk the dependency graph and generate appropriate build-system rules encoding these dependencies. 

Hacky, but in practice it works well.

This is obviously unstable in several different ways, and it would be nice if one of the outputs of the Rust toolchain build process would be a formal dependency graph represented in some machine-readable way (eg as json, perhaps extracted from the Cargo metadata).

(Or see below for another idea.)

3 is very tricky, simply because of the non-local nature.

Allocators need to be genuinely global - all Rust code co-existing in a process must use the same allocator, regardless of how it relates to each other in the dependency graph. The standard approach might be to define some weak symbols for the allocator which can be overridden once. Multiple attempts to override should fail (typically with duplicate symbol linker errors). Rust has a more ergonomic approach to this, but it relies on `rustc` doing the final link. We'd like to find a mechanism which still allows rustc to give good error messages when it does the link, but is still correct if something else is linking.

> Have rustc tell you exactly what you need to link and have this rustc invocation generate any necessary support files like the allocator shim.

(The allocator shim is gone now, right?)

To be honest, I haven't fully grokked the panic-handling side of things. I'm not completely sure what the effect of specifying `--panic=abort`/`unwind` is, and what the implications of different crates being compiled with different panic handling modes is, or what the constraints are. But
> for example only one of panic_abort and panic_unwind may be linked and one of them must be linked if libstd is used
In practice it looks like it's actually fine to include both, since they have non-conflicting symbols. But it would be preferable to have just one.

## link-time magic

Rustc has a few pieces of link-time magic (that I'm aware of, but I'm certainly missing things and getting details wrong):
- allocator shim generation
- panic-handler selection
- choosing a special dylib for shim injection

As mentioned above, ideally we could avoid having an allocator shim at all. But if we must have one, it would be better to have a way for rustc to generate it as a real rlib, rather than slipping it into the crate as a side-effect - eg `rustc --crate-type allocator-shim --emit link=.../liballocshim.rlib <other options>`. That way we can explicitly have a rule to generate a shim and then depend on it like normal.

As I mentioned above, I don't fully understand all the constraints around panic-handlers, but it definitely falls into the link-time magic category.

And the "magic dylib" stuff makes dylibs very non-uniform and in effect the specific artifact that gets produced depends on, uh, the specific order the dependency graph is walked? Actually now that I write this, I don't really understand this mechanism either. But it also falls under the category of "dynamic linking / shared objects are not well supported".


## (Separate binary builds from links)

This is a bit of an aside, but being able to build a `--crate-type bin`'s rust code separately from linking it seems closely related to all this, and an independently valuable thing to have.

## staticlib link step

> By the way I forgot if I asked this already, but would a rustc command to bundle multiple rlibs into a single staticlib without requiring any rust source for this staticlib work? So something like `rustc --crate-type staticlib -Zlink-libs liba.rlib libb.rlib -o libcombined.a`. This should allow diamond dependencies I think. You could run it just before invoking the system linker.

This is awkward for a few reasons:
- one is @adetaylor's concern, that we'd like to avoid having to do O(dep graph) operations at link time, if we can avoid it. The link itself it unavoidably that scale, but we'd still like to avoid more
- secondly, we'd need to apply this to all "final" links, but *only* final links. That is, we have cases where there are sometimes intermediate linked artifacts (eg shared libraries) where we may or may not want to include the output of this "rustc staticlib" step
- thirdly, it puts a requirement on all link steps - ie, we'd need to add a Rust special case to C++, Ocaml, Haskell, ... links - wherever we do a link and there's a chance that Rust could end up somewhere in the dependency graph
- (thin archives avoid the IO overhead, but they're not universally available)

I'm not saying we couldn't make it work, but it wouldn't be straightforward. And we're getting quite far from "rust objects are just ordinary objects" (ie, the "make RIIR easy" case I mentioned above).

Since Buck (v2) supports some level of dynamic dependencies, we could have something like `--emit sysroot-deps=...` for each crate target which would emit the specific sysroot libraries in some form (and maybe shims, etc) which that crate requires. This would have to be abstracted rather than, eg, paths to objects (eg emit `std,core,alloc` rather than `/usr/local/lib/rust/.../libstd-abc123.rlib`) so we can map these to actual target names. That would still allow `rustc` to retain and export knowledge of the internal dependencies on sysroot while allowing us to put them to use them as formal dependencies.

That said, dynamic dependencies are uncommon among build systems - I'm not sure Bazel or GN/ninja or even plain `make` could support this, for example. Buck v1 definitely can't.

> On a side note: The current compilation model of rustc hasn't been well thought out. It consists of a bunch of features bolted on top of each other creating a bit of a mess. I feel like we should first document the current compilation model and have a unifying vision for a better compilation model before we can think about how to incrementally improve (or maybe even rewrite from scratch) the compilation model of rustc to prevent painting ourself into a corner with suboptimal compilation performance while at the same time having bad integration with the C compilation model and permanently bad dylib support

Very much agreed. While we have a bunch of practical motivations and specific goals, along with point fixes for specific problems, I'd love it if it also helps contribute to a "how can we systematically fix Rust's compilation model" conversation. There are a *ton* of engineering tradeoffs which need to be taken into consideration, and there's going to be a fair amount of give and take, but there's a lot of scope for improvement.

**Comment by bjorn3** [comment]
Thanks for the extensive writeup! A couple of notes:

> (The allocator shim is gone now, right?)

Not yet and even with my PR it will only be gone if `#[global_alloctator]` is used.

> In practice it looks like it's actually fine to include both, since they have non-conflicting symbols.

They have conflicting symbols. It may just not look like it because one uses `#[rustc_std_internal_symbol]` which uses an unmangled name while the other uses a lang item where rustc chooses the same symbol name as the other crate.

> I'm not completely sure what the effect of specifying --panic=abort/unwind is, and what the implications of different crates being compiled with different panic handling modes is, or what the constraints are.

If any crate uses -Cpanic=abort, panic_abort must be linked to avoid UB. Otherwise both are technically fine, but panic_unwind is preferd as otherwise compiling object files with unwinding support is kind of pointless.

**Comment by jsgf** [comment]
Thanks for the clarifications.

> They have conflicting symbols. It may just not look like it because one uses #[rustc_std_internal_symbol] which uses an unmangled name while the other uses a lang item where rustc chooses the same symbol name as the other crate.

Right, we'll need to look into that.

> If any crate uses -Cpanic=abort, panic_abort must be linked to avoid UB. Otherwise both are technically fine, but panic_unwind is preferd as otherwise compiling object files with unwinding support is kind of pointless.

I see - so we should consider a `-Cpanic=abort` as a build-wide config option - ie, something you specify on the top-level target (ie binary) and propagate to all its dependencies.

**Comment by jsgf** [comment]
Oh, I meant to add:

> Such a feature would likely require every crate to be compiled with it. For as long as we don't have stable -Zbuild-std that means the standard library would have to be compiled with this feature, but that would break build systems depending on rlibs containing useful object files

@mhammerly has prototyped generating build rules for the sysroot libraries (using [reindeer](https://github.com/facebookincubator/reindeer)) so that they're built on the fly along with everything else. As far as the build system is concerned, they're just some slightly strange third-party libraries.

So something that requires standard library to be built with specific flags is not necessarily a blocker (though it does mean yet more stuff to stabilize if its required for a stable implementation of all this).

**Comment by bjorn3** [comment]
The allocator shim is an implementation detail and would likely be removed in favor of weak symbols in libstd if weak symbols were universally supported across all platforms. https://github.com/rust-lang/rust/pull/86844 in combination with a single mandatory `#[global_allocator]` somewhere would be better I think. If necessary this could be automatically provided by the build system in the form of compiling `#[global_allocator] static ALLOC: std::alloc::System = std::alloc::System;`.

**Comment by hlopko** [comment]
As long as there is a stabilized way of wiring Rust with the system allocator I think our use case is covered. I don't see any reason why your solution wouldn't work for us, so thumbs up :)

**Comment by jsgf** [comment]
> The allocator shim is an implementation detail and would likely be removed in favor of weak symbols in libstd if weak symbols were universally supported across all platforms

As an implementation detail, it would seem fine to use in the cases where weak symbols are not available, though it would obviously nicer to have a single mechanism rather than having to support two. I guess it depends on how common platforms without weak symbol support are? Oh, I guess that includes obscure little systems like Windows. Though maybe you could do something functionally equivalent at the library level with the library search order and/or `/DEFAULTLIB:`.

**Comment by jsgf** [comment]
Oh, another thing I forgot to mention -

The `libstd` build process copies parts of the llvm runtime into itself - in particular the various sanitizer runtimes. This leads to duplicate symbols if you're linking with sanitized C/C++ code. Instead `rustc` should link with the llvm-provided libraries when the sanitizer options are enabled. (This assumes the rustc llvm sanitizer is compatible with the clang llvm sanitizer. We guarantee this by compiling `rustc` with the same `llvm` that we use for `clang`, which also makes sure that LTO works.)

**Comment by durin42** [comment]
That's also what we're moving towards (rapidly) for rustc/clang compatibility. We'll probably start working on ThinLTO support in the bazel rules for Rust in the next week or two.

The duplicate symbols for mixed C++/Rust sanitizer builds hadn't occurred to me. I'll file an internal tracking bug about that, thanks.

**Comment by dureuill** [comment]
Hello and apologies for bumping this issue,

We're hitting this big time as we try to scale to multiple independent [CXX](cxx.rs) libraries in our CMake-based build system.

Lack of stability promises apart, are there known immediate issues to linking directly the rlibs with e.g. clang++ or g++?

As a way to avoid looking for the system rlibs, I'm experimenting with building a `libbaserust.a` through an empty cargo project configured to produce a `staticlib`.

This allows to switch from:

```
g++ test.cpp -std=c++14 target/debug/build/rs2cpp-72074e3e9d2e0779/out/librsffi.a target/debug/librs2cpp.rlib target/debug/deps/*.rlib /home/tetrane/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/*.rlib -pthread -ldl
```

to:

```
g++ test.cpp -std=c++14 target/debug/build/rs2cpp-72074e3e9d2e0779/out/librsffi.a target/debug/librs2cpp.rlib target/debug/deps/*.rlib base-rust/target/debug/libbaserust.a -pthread -ldl
```

Is that worse or the same as the full rlib based approach?

Thank you for reading

**Comment by bsilver8192** [comment]
I've got it working with bazelbuild/rules_rust#1350. That's linking the rlibs, plus a .c file to handle the allocator shims. Seems to be pretty robust for Linux without sanitizers, but it does require Rust-version-specific tweaks to the ordering for the system rlibs and contents of the .c file.

bazelbuild/rules_rust#1238 has some more discussion. Somewhere in the discussion related to that (can't find it now) there was mention of several other projects using similar approaches.

**Comment by dureuill** [comment]
Ah I don't seem to be encountering the issue with allocators in the `libbaserust.a` version.

I understand that it is an issue when linking with only the rlibs?

**Comment by bsilver8192** [comment]
Hmmm, not sure. Maybe the `staticlib` includes those symbols? But I'm not sure if it's guaranteed to include all of them or just the ones used by the Rust code in it.

**Comment by dureuill** [comment]
Just to make sure we're talking about the same issue, when attempting to build with the following command line:

```
g++ test.cpp -std=c++14 target/debug/build/rs2cpp-72074e3e9d2e0779/out/librsffi.a target/debug/librs2cpp.rlib target/debug/deps/*.rlib /home/tetrane/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/*.rlib -pthread -ldl
```

I get linker errors like:

```
/usr/bin/ld: /home/tetrane/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc-b8438dc0bcbbcc08.rlib(alloc-b8438dc0bcbbcc08.alloc.bf51ef42-cgu.0.rcgu.o): in function `alloc::alloc::dealloc':
/rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e//library/alloc/src/alloc.rs:107: undefined reference to `__rust_dealloc'
/usr/bin/ld: /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e//library/alloc/src/alloc.rs:107: undefined reference to `__rust_dealloc'
/usr/bin/ld: /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e//library/alloc/src/alloc.rs:107: undefined reference to `__rust_dealloc'
```

Is that the issue at hand with the allocators?

If so, when making an additional empty rust project `rustbase` and setting it to compile with `staticlib`, and lastly using this static lib on top of the rlibs from my actual project:

```
g++ test.cpp -std=c++14 target/debug/build/rs2cpp-72074e3e9d2e0779/out/librsffi.a target/debug/librs2cpp.rlib target/debug/deps/*.rlib base-rust/target/debug/libbaserust.a -pthread -ldl
```

then the link is successful. My (limited) understanding is that whatever "base rust setup" needed must be included even in the simplest staticlib. This behavior might be OS and toolchain dependent, though, and could also have "blind spot".

**Comment by bsilver8192** [comment]
Yep, that's the allocator shims. There's more than just `__rust_alloc` and `__rust_dealloc` that you might hit with more complex code, and I'm not sure if your `staticlib` will include those or not.

**Comment by dureuill** [comment]
Could you maybe share any way of triggering the missing functions from standard code, so that I could test if the staticlib include these?

`nm` on the staticlib tells me that it contains ("T" type symbols) `__rust_alloc`, `__rust_alloc_error_handler`, `__rust_alloc_error_handler_should_panic`, `__rust_alloc_zeroed`, `__rust_realloc`, `rust_oom`, `__rdl_oom` and `__rust_dealloc`

**Comment by hlopko** [comment]
This is the file with definitions that we use with the latest rustc: https://github.com/google/crubit/blob/main/common/rust_allocator_shims.c. 

This is where rustc declares them: https://stdrs.dev/nightly/x86_64-unknown-linux-gnu/alloc/alloc/index.html.

I'd expect that `staticlib` defines all of these. Still, using a mostly empty `staticlib` and then `.rlibs` is not officially supported mode, just as it is unsupported to define our own allocator shims :)

**Comment by bjorn3** [comment]
I got a PR for removing the need for the allocator shim when using `#[global_allocator]`: https://github.com/rust-lang/rust/pull/86844 I need to update it some day.

**Comment by dureuill** [comment]
FWIW, if you're able to relax a bit the requirements of only compiling each lib once, we got a setup working with shared libraries.

Each Rust project is configured to generate a `staticlib` (`rustlib.a`), then this `rustlib.a` is linked with a C++ shim (since we're using [cxx](https://cxx.rs) we have to compile the C++ part of the bridge anyway, so that's our shim) and into a shared library `lib.so`.

We pass `-Bsymbolic` to the linker when building the shared libraries, so that each of them lookup symbols internally first, then globally if missing, so as to limit the impact of diamond dependencies.

This works rather well with cxx, because most of the calls will be through the bridge part, which is typically unique to each `.so`. We just have to ensure that we use the same version of cxx (and of rust, probably) across all `.so` when using "vocabulary types" of cxx, such as `rust::str` from the C++, because it will pick the corresponding symbols (ctor, dtor, ...) in one of the `.so`, and then pass the resulting object to the `.so` corresponding to the bridge.

This caveat apart, I believe that this setup should minimize duplicated symbols issues and does not rely on unstable behavior, at the cost of using shared libraries (no more single executable, duplicated code in each shared library for a pretty hefty size).

(I realize that relaxing these requirements is probably not an option for OP, but I'm posting this assuming that others might have different requirements, and in the hope that my experience might be useful)

**Comment by petrochenkov** [comment]
I wanted to confirm one thing.
Scenarios in which people want to use rlibs as "regular static libraries" also assume `-Z link-native-libraries=false`, right?
I.e. the linking of libraries is managed entirely by some outer build system rather than rustc.

In https://github.com/rust-lang/rust/issues/99429 we may want to change representation of native libraries bundled into `.rlib` files, and it would be easier to do if we knew that third parties do not need to rely on the currently used representation (all individual object files from native libraries are copied to the rlib).

**Comment by bsilver8192** [comment]
> I wanted to confirm one thing. Scenarios in which people want to use rlibs as "regular static libraries" also assume `-Z link-native-libraries=false`, right? I.e. the linking of libraries is managed entirely by some outer build system rather than rustc.

+1 for my use case being like this, I want to manage those via Bazel.

**Comment by jsgf** [comment]
@petrochenkov Yes, that's what I'd expect.

> I wanted to confirm one thing.
Scenarios in which people want to use rlibs as "regular static libraries" also assume -Z link-native-libraries=false, right?
I.e. the linking of libraries is managed entirely by some outer build system rather than rustc.

In https://github.com/rust-lang/rust/issues/99429 we may want to change representation of native libraries bundled into .rlib files, and it would be easier to do if we knew that third parties do not need to rely on the currently used representation (all individual object files from native libraries are copied to the rlib).

UPD: Hmm, it looks like -Z link-native-libraries is not actually respected when bundling native libraries into rlibs, not sure whether it's intentional or not.

**Comment by pcwalton** [comment]
Hello! I've posted [a pre-RFC for minimal stabilization of the `rlib` format](https://internals.rust-lang.org/t/pre-rfc-stabilize-a-version-of-the-rlib-format/17558) on internals. Please feel free to comment there. Thanks!

**Comment by bjorn3** [comment]
As of https://github.com/rust-lang/rust/pull/86844 (scheduled for the 1.71 release) if you are directly linking the rlibs of the standard library rather than letting rustc handle linking, you will now need to define a static named `__rust_no_alloc_shim_is_unstable` which is at least 1 byte big. In addition if you are using `#[global_allocator]`, you must stop defining `__rust_alloc`, `__rust_dealloc`, `__rust_realloc` and `__rust_alloc_zeroed` as they are now directly defined by the `#[global_allocator]` expansion rather than as part of the allocator shim. If you are using the default allocator in libstd you will need to keep defining them though.

**Comment by Lupus** [comment]
Is there somewhat future-proof workaround available as part of some open-source project that one could leverage maybe? There are some recipes in this discussion, but it's not clear how one could reconstruct the required "ugly hacks" to get going while we waiting for the right solution to make it to the upstream.

I'm building some bindings from Rust to OCaml, everything worked great until I tried to link two such bindings libraries in one binary, which lead me here with a bunch of linker errors at hand...

**Comment by keith** [comment]
Folks using bazel and [rules_rust](https://github.com/bazelbuild/rules_rust/) workaround this today, so you can trace back that code or an example link command there to see the result

**Comment by danakj** [comment]
The current work around is to build rust rlibs, not staticlibs, and link those as you would .a files

**Comment by Lupus** [comment]
Is compilation of an empty crate as a staticlib still a decent approach to not hunt individual stdlib rlibs? Bazel rules around that are quite wordy as it seems...

Other than that, one needs to parse cargo manifest, build dependency graph, build a list of rlibs required for particular crate, pass that list to the linker?

**Comment by durin42** [comment]
AIUI, it's a decent approach if-and-only-if you can consolidate all your Rust bits into a single unified target. Otherwise you run the risk of pulling stuff in twice, which either bloats your binary or causes linker errors depending on how you do it.

**Comment by tgross35** [comment]
@pcwalton did you ever move forward from pre-rfc to rfc with that?

I don't think I've seen this yet (it's a long thread...) but could `rustc` maybe gain the ability to turn a `rlib` into a `.a`? It would be a pretty fast operation that would let us do something before we get to the point of stabilizing rlib (still +1 for the RFC of course) and without needing to figure out a non-rlib distribution of std.

**Comment by bjorn3** [comment]
We already have the staticlib crate type for bundling everything into a single `.a` file. Turning individual rlibs into `.a` files won't work as it would either duplicate the allocator shim and such between every such `.a` or or omit them from all and produce unlinkable `.a` files. One idea I have is to produce a new crate type which is to `staticlib` as `dylib` is to `cdylib`. It would act like a `dylib` with respect to producing the allocator shim and bundling multiple crates, but produce a `.a` file instead of a `.so` file and add the necessary crate metadata to allow consuming it like a regular crate from rustc. See also the end of https://github.com/rust-lang/rust/issues/111594#issuecomment-1550087206

**Comment by adamncasey** [comment]
Myself and a few others at my company would like to revive this effort and contribute what's needed to get this over the line.

The end goals we would like to see are:
1. `rustc` documents and supports ([here?](https://doc.rust-lang.org/reference/linkage.html#mixed-rust-and-foreign-codebases)) a build + non-rustc rlib link model as a first-class workflow. Likely with some constraints to keep support feasible.
2. Tests in the rust repo ensure this workflow doesn't regress (at minimum for `x86_64-unknown-linux-gnu`)
3. Via the above, for C++ applications being built in a mono-repo/hermetic-build-like environment to be able to link against a library built in rust without any changes to its link process.

From digging a bit into how bazel/buck2 work, and from some conversations at conferences, this workflow is important at a couple of large organisations which are big in the Rust space.
In particular this workflow feels critical for integrating Rust gradually into codebases traditionally built with just C and C++.

I'm updating this issue primarily to get an indication from maintainers of how likely it is that this support could be formalised, but also to figure out how many others are interested in this. Reactions/feedback are appreciated.

On our side the next steps are to re-publish/re-open the RFC @pcwalton wrote to collect feedback. Based on the feedback received we will look at implementation/tests/documentation afterwards.

**Comment by bjorn3** [comment]
I don't think we can reasonably support this with the direct rlib linking path until EII (externally implementable items) is fully supported and we have switched the global allocator and alloc error handler over to using it. That is currently blocked on Windows weak linking being a bit of a pain. I believe there are some ideas to make it work, but it will take time.

After that I think one option is specifying exactly which configurations and features are supported and which aren't (eg bundled static libs (as is the default when linking static libs) can't work without rustc unbundling them right before linking) and then formally blessing it. 

The other option is to introduce a rust dylib counterpart to static libraries (maybe rstaticlib?). Something that can bundle multiple rlibs and runs all global consistency checks (no conflicting definitions of EIIs, no (different versions of a) crate getting linked twice, ...), while still being usable as regular rust dependency just like rlibs and rust dylibs. You could then have one such rstaticlib for the standard library + crate that defines a global allocator if you want to override it and after that one rstaticlib each for every user crate that is linkable as just another C static library in the end, while still benefitting from the consistency checks in rustc and having full support for all features of rust, including those directly linking fundamentally rlibs can't offer like bundled staticlibs.

ps: Which RFC are you referring to? I can't find any closed RFCs on rust-lang/rust by pcwalton that could be relevant.

**Comment by adamncasey** [comment]
Thanks for the quick reply. Sorry I meant this pre-RFC: https://internals.rust-lang.org/t/pre-rfc-stabilize-a-version-of-the-rlib-format/17558

I agree that the combination of EII and this needs some thought. I've got some reading to do.

For those already depending on this workflow today - do you have any initial plans for handling EIIs in practice?
